worlds/zork_grand_inquisitor/data_funcs.py

```diff
@@ -0,0 +1,247 @@
+        if data.statemap_keys is not None:
+            for key in data.statemap_keys:
+                mapping[key] = item
+
+    return mapping
+
+
+def location_groups() -> Dict[str, Set[str]]:
+    groups: Dict[str, Set[str]] = dict()
+
+    tag: ZorkGrandInquisitorTags
+    for tag in ZorkGrandInquisitorTags:
+        groups[tag.value] = set()
+
+    location: ZorkGrandInquisitorLocations
+    data: ZorkGrandInquisitorLocationData
+    for location, data in location_data.items():
+        if data.tags is not None:
+            for tag in data.tags:
+                groups[tag.value].add(location.value)
+
+    return {k: v for k, v in groups.items() if len(v)}
+
+
+def locations_by_region(include_deathsanity: bool = False) -> Dict[
+    ZorkGrandInquisitorRegions, Set[ZorkGrandInquisitorLocations]
+]:
+    mapping: Dict[ZorkGrandInquisitorRegions, Set[ZorkGrandInquisitorLocations]] = dict()
+
+    region: ZorkGrandInquisitorRegions
+    for region in ZorkGrandInquisitorRegions:
+        mapping[region] = set()
+
+    location: ZorkGrandInquisitorLocations
+    data: ZorkGrandInquisitorLocationData
+    for location, data in location_data.items():
+        if include_deathsanity is False and ZorkGrandInquisitorTags.DEATHSANITY in (
+            data.tags or tuple()
+        ):
+            continue
+
+        mapping[data.region].add(location)
+
+    return mapping
+
+
+def locations_with_tag(tag: ZorkGrandInquisitorTags) -> Set[ZorkGrandInquisitorLocations]:
+    location: ZorkGrandInquisitorLocations
+    data: ZorkGrandInquisitorLocationData
+
+    return {location for location, data in location_data.items() if data.tags is not None and tag in data.tags}
+
+
+def location_access_rule_for(location: ZorkGrandInquisitorLocations, player: int) -> str:
+    data: ZorkGrandInquisitorLocationData = location_data[location]
+
+    if data.requirements is None:
+        return "lambda state: True"
+
+    lambda_string: str = "lambda state: "
+
+    i: int
+    requirement: Union[
+        Tuple[
+            Union[
+                ZorkGrandInquisitorEvents,
+                ZorkGrandInquisitorItems,
+            ],
+            ...,
+        ],
+        ZorkGrandInquisitorEvents,
+        ZorkGrandInquisitorItems
+    ]
+
+    for i, requirement in enumerate(data.requirements):
+        if isinstance(requirement, tuple):
+            lambda_string += "("
+
+            ii: int
+            sub_requirement: Union[ZorkGrandInquisitorEvents, ZorkGrandInquisitorItems]
+            for ii, sub_requirement in enumerate(requirement):
+                lambda_string += f"state.has(\"{sub_requirement.value}\", {player})"
+
+                if ii < len(requirement) - 1:
+                    lambda_string += " or "
+
+            lambda_string += ")"
+        else:
+            lambda_string += f"state.has(\"{requirement.value}\", {player})"
+
+        if i < len(data.requirements) - 1:
+            lambda_string += " and "
+
+    return lambda_string
```

Comment:
Context here is that I went with a bottom-up approach from pure data. My data is rich but pure in the sense that it holds nothing functional in it; only literals and enum items.

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/7ca8be5f-07d0-43d3-963d-fca1f18f257f)

The majority of the functionality of my world (both AP-side and client-side) leverages this data. It's way more useful and versatile than typical tables worlds use. For rules, I know that hand-typed lambdas are more common (and more performant than evaluating at runtime, not going to try and pretend otherwise), but I find it extremely error-prone and messy. Here, I can unit test 2 functions and have peace of mind that everything will work as intended from a simple data structure.

Reply:
That being said, it's not that I specifically WANT to do lambda strings with evaluation. I just haven't found a better way that works with the moderate level of rule complexity I have.

If someone can show me a better way to do it from the data I have, I am 100% in.

Reply:
have you perused https://docs.python.org/3/library/ast.html yet?